### PR TITLE
Use public hostnames (with SSL)to provide transparent networking between host machine and container environment

### PR DESCRIPTION
**Pull request by elephant3**
## Problem
Services can communicate internally, but there was no way of unified hostname that was available inside docker network and on host machine.

## Solution
Use a set of hostnames which resolve to `127.0.0.1`  by our nameservers and pass it as an `extra_hosts` parameter to docker compose services.

### CODE REVIEW

**Comment by elitan** [comment]
My browser forces `https` for the dahshboard, console, and mailhog. Even though we're using http. Maybe we should use `localhost` for the dashboard, console, and mailhog instead.

File: .github/cert.sh [hunk 1]
```diff
@@ -0,0 +1,15 @@
+#!/bin/sh
+
```
Comment (dbarrosop) [inline .github/cert.sh:2]: I'd suggest starting with `set -euo pipefail`

File: nhost/compose/config.go [hunk 2]
```diff
@@ -33,13 +30,14 @@ const (
...
-	proxyPort   = 1337
+	graphqlPort  = 8080
+	dbPort       = 5432
+	proxyPort    = 1337
+	proxySSLPort = 443
 	// --
 
 	// default docker images
-	svcDashboardDefaultImage = "nhost/dashboard:0.9.10"
+	svcDashboardDefaultImage = "dashboard" // TODO: update image name before release
```
Comment (elephant3) [inline nhost/compose/config.go:40]: # IMPORTANT: update this value before merging
This is set to `dashboard` for easier local testing with changes from https://github.com/nhost/nhost/pull/1467

Build docker image with changes from the mentioned PR above with: `docker build -f dashboard/Dockerfile -t dashboard . `

File: internal/ssl/.ssl/fullchain.pem [hunk 3]
```diff
@@ -0,0 +1,90 @@
+-----BEGIN CERTIFICATE-----
```
Comment (dbarrosop) [inline internal/ssl/.ssl/fullchain.pem:1]: are these "good" certificates? I'd suggest using dummy certs here and generate good ones on the fly at build time in the CI.

Comment (elephant3) [reply internal/ssl/.ssl/fullchain.pem:1]: Yeah, this one is real. I'll replace it with self-signed and provide a simple shell script that generates a real one for internal usage and testing.

File: cmd/dev.go [hunk 4]
```diff
@@ -224,14 +250,48 @@ var devCmd = &cobra.Command{
...
+func checkHostnames() error {
+	hostnames := []string{compose.HostLocalDashboardNhostRun, compose.HostLocalGraphqlNhostRun, compose.HostLocalAuthNhostRun, compose.HostLocalStorageNhostRun, compose.HostLocalFunctionsNhostRun, compose.HostLocalMailNhostRun}
+
+	for _, hostname := range hostnames {
+		_, err := net.LookupIP(compose.HostLocalDashboardNhostRun)
+		if err != nil {
+			logger.DEBUG = false
+			fmt.Println(fmt.Sprintf(`Failed to resolve '%s' hostname
+
+Make sure that you have internet connection.
```
Comment (nunopato) [inline cmd/dev.go:262]: Please make sure you have an internet connection and try again.
```suggestion
Make sure that you have internet connection.
```

File: cmd/dev.go [hunk 5]
```diff
@@ -224,14 +250,48 @@ var devCmd = &cobra.Command{
...
+
+If you want to use CLI offline, make sure to add the following lines in /etc/hosts:
```
Comment (nunopato) [inline cmd/dev.go:264]:
```suggestion
If you don't have an internet connection, please add the following lines to /etc/hosts:
```

File: cmd/dev.go (hunk 4 above)
Comment (nunopato) [inline cmd/dev.go:262]:
```suggestion
Please make sure you have an internet connection and try again.
```

File: cmd/dev.go [hunk 6]
```diff
@@ -224,14 +250,48 @@ var devCmd = &cobra.Command{
 	},
 }
 
+func checkHostnames() error {
+	hostnames := []string{compose.HostLocalDashboardNhostRun, compose.HostLocalGraphqlNhostRun, compose.HostLocalAuthNhostRun, compose.HostLocalStorageNhostRun, compose.HostLocalFunctionsNhostRun, compose.HostLocalMailNhostRun}
+
+	for _, hostname := range hostnames {
+		_, err := net.LookupIP(compose.HostLocalDashboardNhostRun)
```
Comment (nunopato) [inline cmd/dev.go:257]: do you mean to pass `hostname` here?

File: cmd/dev.go [hunk 7]
```diff
@@ -224,14 +250,48 @@ var devCmd = &cobra.Command{
...
+%s
+`, hostname, offlineConfigForSslHostnames(hostnames)))
+			return fmt.Errorf("failed to resolve '%s' hostname: %v", hostname, err)
+		}
+	}
+
+	return nil
+}
+
+func offlineConfigForSslHostnames(hostnames []string) string {
```
Comment (nunopato) [inline cmd/dev.go:275]:
```suggestion
func offlineConfigForSSLHostnames(hostnames []string) string {
```

File: cmd/dev.go [hunk 8]
```diff
@@ -224,14 +250,48 @@ var devCmd = &cobra.Command{
...
+	lines := []string{}
+	for _, hostname := range hostnames {
+		lines = append(lines, fmt.Sprintf("127.0.0.1	%s", hostname))
+	}
+	return strings.Join(lines, "\n")
+}
+
 func getPorts(fs *flag.FlagSet) (*ports.Ports, error) {
-	var proxyPort, dbPort, graphqlPort, hasuraConsolePort, hasuraConsoleApiPort, smtpPort, minioS3Port, mailhogPort, dashboardPort uint32
+	var proxyPort, sslProxyPort, dbPort, graphqlPort, hasuraConsolePort, hasuraConsoleApiPort, smtpPort, minioS3Port uint32
```
Comment (nunopato) [inline cmd/dev.go:284]: I suggest using a declaration block for better readability and renaming `hasuraConsoleApiPort` to `hasuraApiPort`.

File: nhost/compose/config.go [hunk 9]
```diff
@@ -126,9 +78,15 @@ func NewConfig(conf *nhost.Configuration, p *ports.Ports, env []string, gitBranc
 	return &Config{nhostConfig: conf, ports: p, dotenv: env, gitBranch: gitBranch, composeProjectName: projectName}
 }
 
-func (c Config) addLocaldevExtraHost(svc *types.ServiceConfig) *types.ServiceConfig {
+func (c Config) addExtraHosts(svc *types.ServiceConfig) *types.ServiceConfig {
```
Comment (nunopato) [inline nhost/compose/config.go:81]: I am not sure we need to set all hosts for each service if the value for all fields is always the same.

File: nhost/compose/hostnames.go [hunk 10]
```diff
@@ -0,0 +1,84 @@
...
+
+func MinioHostname(port uint32) string {
+	return httpHostnameWithPort(hostMinio, port)
+}
+
+func HasuraConsoleHostname(port uint32) string {
+	return httpHostnameWithPort(hostHasuraConsole, port)
+}
+
+func HasuraAPIHostname(port uint32) string {
```
Comment (nunopato) [inline nhost/compose/hostnames.go:36]: do you mean hasura metadata api here?

File: cmd/dev.go (hunk 6 above)
Comment (elephant3) [reply cmd/dev.go]: good catch 😆

File: nhost/compose/hostnames.go (hunk 10 above)
Comment (elephant3) [reply nhost/compose/hostnames.go]: Actually I realized this method isn't needed anymore, updated the code

**Comment by fungilation** [comment]
I'll lump a bug report in this PR, to test for this in whether this PR resolves. cc @elitan @nunopato 

I have a event trigger, `users-post-insert`, which trigger on any insert into auth.users table. On local nhost-cli however, hasura event trigger fail to connect with the local function (http://localhost:1337/v1/functions/users-post-insert). Hasura event trigger error response:

```
{
  "data": {
    "message": "{'message':'Connection failure','request':{'host':'localhost','method':'POST','path':'/v1/functions/users-post-insert','port':1337,'queryString':'','requestHeaders':{'Content-Type':'application/json','User-Agent':'hasura-graphql-engine/v2.15.2','X-B3-ParentSpanId':'a74a578cb7699c6f','X-B3-SpanId':'b960b8a0cf1c6058','X-B3-TraceId':'eff36af96c718b6beadfbcf8a56cee46','nhost-webhook-secret':'nhost-webhook-secret'},'responseTimeout':'ResponseTimeoutMicro 60000000','secure':false},'type':'http_exception'}"
  },
  "type": "client_error",
  "version": "2"
}
```

http://localhost:1337/v1/functions/users-post-insert however can be reached on my computer by browser. So, networking timeout only from local hasura -> functions. In the past, I workarounded this by using a lan IP instead of `localhost`, but now with regions config, anything other than localhost is no longer workable with nhost cli.

**Review by nunopato — APPROVED** [review]

**Pull request merged**